File: Codes/1027-longest-arithmetic-subsequence.py
'''
Author: LetMeFly
Date: 2023-04-22 09:42:46
LastEditors: LetMeFly
LastEditTime: 2023-04-22 09:49:42
'''
from typing import List

class Solution:
    def longestArithSeqLength(self, nums: List[int]) -> int:
        ans = 2
        diff = max(nums) - min(nums)
        for d in range(-diff, diff + 1):
            mp = dict()
            for num in nums:
                if num - d in mp:
                    thisAns = mp[num - d] + 1
                else:
                    thisAns = 1
                mp[num] = thisAns
                ans = max(ans, thisAns)
        return ans

File: Codes/1027-longest-arithmetic-subsequence.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-04-22 09:42:41
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-04-22 09:47:09
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    int longestArithSeqLength(vector<int>& nums) {
        int ans = 2;
        auto minmax = minmax_element(nums.begin(), nums.end());
        int diff = *minmax.second - *minmax.first;
        for (int d = -diff; d <= diff; d++) {  // 要从-diff开始
            unordered_map<int, int> ma;
            for (int num : nums) {
                int thisAns;
                if (ma.count(num - d)) {
                    thisAns = ma[num - d] + 1;
                }
                else {
                    thisAns = 1;
                }
                ma[num] = thisAns;
                ans = max(ans, thisAns);
            }
        }
        return ans;
    }
};